squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+
+    Notes
+    -----
+    If some genes in :attr:`anndata.AnnData.uns` ``['sepal_score']`` are `NaN`,
+    consider re-running the function with increased `n_iter`.
+    """
+    _assert_connectivity_key(adata, connectivity_key)
+    _assert_spatial_basis(adata, key=spatial_key)
+
+    spatial = adata.obsm[spatial_key].astype(np.float_)
+
+    if genes is None:
+        if "highly_variable" in adata.var.columns:
+            genes = adata[:, adata.var.highly_variable.values].var_names.values
+        else:
+            genes = adata.var_names.values
+    genes = _assert_non_empty_sequence(genes, name="genes")
+
+    n_jobs = _get_n_cores(n_jobs)
+
+    vals = _get_obs_rep(adata[:, genes], use_raw=use_raw, layer=layer)
+    vals = _resolve_vals(vals)
+
+    g = adata.obsp[connectivity_key].copy()
+
+    max_n = np.diff(g.indptr).max()
+    if max_n != max_nbrs:
+        logg.warning(f"Found node with # neighbors == {max_n}, bu expected `max_nbrs` == {max_nbrs}.")
+
+    # get saturated/unsaturated nodes
+    sat, sat_idx, unsat, unsat_idx = _compute_idxs(g, spatial, max_nbrs, "l1")
+
+    start = logg.info(f"Calculating sepal score for `{len(genes)}` genes using `{n_jobs}` core(s)")
+
+    score = parallelize(
+        _score_helper,
+        collection=np.arange(len(genes)),
+        extractor=np.hstack,
+        use_ixs=False,
+        n_jobs=n_jobs,
+        backend=backend,
+        show_progress_bar=show_progress_bar,
+    )(
+        vals=vals,
+        max_n=max_nbrs,
+        n_iter=n_iter,
+        sat=sat,
+        sat_idx=sat_idx,
+        unsat=unsat,
+        unsat_idx=unsat_idx,
+        dt=dt,
+        thres=thres,
+    )
+
+    sepal_score = pd.DataFrame(score, index=genes, columns=["sepal_score"])
+    if sepal_score["sepal_score"].isna().any():
+        logg.info("Found NaN in sepal scores, consider increase `n_iter` to a higher value.")
+    sepal_score.sort_values(by="sepal_score", ascending=False, inplace=True)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return sepal_score
+
+    _save_data(adata, attr="uns", key="sepal_score", data=sepal_score, time=start)
+
+
+def _score_helper(
+    ixs: Sequence[int],
+    vals: Union[spmatrix, np.ndarray],
+    max_n: int,
+    n_iter: int,
+    sat: np.ndarray,
+    sat_idx: np.ndarray,
+    unsat: np.ndarray,
+    unsat_idx: np.ndarray,
+    dt: np.float_,
+    thres: np.float_,
+    queue: Optional[SigQueue] = None,
+) -> np.ndarray:
+
+    score = []
+    if max_n == 4:
+        fun = _laplacian_rect
+    elif max_n == 6:
+        fun = _laplacian_hex
+    else:
+        NotImplementedError("Laplacian for `max_nbrs`= {max_n} is not yet implemented.")
+    for i in ixs:
+        conc = vals[:, i].copy()
+        time_iter = _diffusion(
+            conc, fun, n_iter, sat, sat_idx, sat.shape[0], unsat, unsat_idx, conc.shape[0], dt=dt, thres=thres
+        )
+        if time_iter.shape[0] != 0:
```

Comment:
I'd make `_diffusion` return a float since you're just accessing the 1st elem of the entropy array (if present).
If the array is empty, just return `np.nan` from `_diffusion`. It will save you again some memory + `if` in this loop.

Reply:
> a float since you're just accessing the 1st elem of the entropy array (if present).

it's not the first element but the first array index. I then use the shape of the index to count the number of steps it required, and multiply it by dt to get the final "time"

Reply:
Ok, I get it now. I'd still do the below since it's a bit cleaner I think:
```python
# _diffusion
tmp = np.nonzero(entropy_arr <= thresh)[0]
return tmp[0] if len(tmp) else np.nan  # or tmp.shape[0] if tmp.shape[0] else np.nan
# _score_helper
time_iter = _diffusion(...)
scores.append(time_iter * dt)
```

Reply:
Done.